### ISSUE TITLE
wrong use of sequence_loss_by_example in ptb_word_lm.py?

### ISSUE THREAD

**Issue by jinghuangzhu** [tensorflow/tensorflow#3375]
Hi, I noticed "sequence_loss_by_example" with default of "average_across_timesteps":

def sequence_loss_by_example(logits, targets, weights,
                             average_across_timesteps=True,
                             softmax_loss_function=None, name=None):

but in pub_word_lm.py:

loss = tf.nn.seq2seq.sequence_loss_by_example(
        [logits],
        [tf.reshape(self._targets, [-1])],
        [tf.ones([batch_size * num_steps])])
    self._cost = cost = tf.reduce_sum(loss) / batch_size

weights sum of batch_size*num_steps was used in "sequence_loss_by_example" to normalize the score, so there is no need to divide loss by batch_size again, right? cost was used for gradient later.

**Comment by concretevitamin** [comment]
@ebrevdo Eugene, can you take a look?

**Comment by ebrevdo** [comment]
Lukasz knows this code better.

**Comment by lukaszkaiser** [comment]
Hi!

I think the use of sequence_loss_by_example in ptb_word_lm in ok, because sequence_loss_by_example averages across time-steps (not batch), but ptb_word_lm uses it with just 1 time-step -- it first puts the time component into the batch dimenstion. That's why it has the reshape, but the lists are all of length 1 -- that's the time dimenstion and that's why you see "batch_size * num_steps" there in weights.

Hope that clarifies it!

Lukasz

**Comment by yl-jiang** [comment]
When i run the following demo:

import tensorflow as tf

A = tf.constant([[0.1,0.2,0.3,0.4],[0.2,0.1,0.4,0.3],[0.4,0.3,0.2,0.1],[0.3,0.2,0.1,0.4],[0.1,0.4,0.3,0.2]], dtype=tf.float32)
B = tf.constant([1, 2, 1, 3, 3], dtype=tf.int32)
w_1 = tf.constant(value=[1,1,1,1,1], dtype=tf.float32)
w_2 = tf.constant(value=[1,2,3,4,5], dtype=tf.float32)

D = tf.contrib.legacy_seq2seq.sequence_loss_by_example([A], [B], [w_1])
D_1 = tf.contrib.legacy_seq2seq.sequence_loss_by_example([A], [B], [w_1], average_across_timesteps=False)
D_2 = tf.contrib.legacy_seq2seq.sequence_loss_by_example([A], [B], [w_2])
D_3 = tf.contrib.legacy_seq2seq.sequence_loss_by_example([A], [B], [w_2], average_across_timesteps=False)

with tf.Session() as sess:
    print(sess.run(D))
    print(sess.run(D_1))
    print(sess.run(D_2))
    print(sess.run(D_3))
import tensorflow as tf

A = tf.constant([[0.1,0.2,0.3,0.4],[0.2,0.1,0.4,0.3],[0.4,0.3,0.2,0.1],[0.3,0.2,0.1,0.4],[0.1,0.4,0.3,0.2]], dtype=tf.float32)
B = tf.constant([1, 2, 1, 3, 3], dtype=tf.int32)
w_1 = tf.constant(value=[1,1,1,1,1], dtype=tf.float32)
w_2 = tf.constant(value=[1,2,3,4,5], dtype=tf.float32)

D = tf.contrib.legacy_seq2seq.sequence_loss_by_example([A], [B], [w_1])
D_1 = tf.contrib.legacy_seq2seq.sequence_loss_by_example([A], [B], [w_1], average_across_timesteps=False)


with tf.Session() as sess:
    print(sess.run(D))
    print(sess.run(D_1))
   

The outputs are as following:

[1.4425355 1.2425355 1.3425356 1.2425356 1.4425356]
[1.4425355 1.2425355 1.3425356 1.2425356 1.4425356]

I found that whether the parm 'average_across_timesteps' is set as  'True' or 'False' the result is the same, why?
